src/execution/collectFields.ts

```diff
@@ -37,16 +42,34 @@ export function collectFields(
   runtimeType: GraphQLObjectType,
   selectionSet: SelectionSetNode,
 ): Map<string, ReadonlyArray<FieldNode>> {
+  const foundFragments: Array<FragmentEntry> = [];
   const fields = new Map();
+  const visited = new Set<string>();
   collectFieldsImpl(
     schema,
     fragments,
     variableValues,
     runtimeType,
     selectionSet,
     fields,
-    new Set(),
+    visited,
+    foundFragments,
   );
+
+  let entry;
+  while ((entry = foundFragments.pop()) !== undefined) {
```

Comment:
We could switch to leveraging `shift()` here if we are concerned with the order these are processed in, my assumption was that order doesn't matter as we come to a consistent view either way

Reply:
Probably better to preserve order, see https://github.com/graphql/graphql-spec/pull/1092

Reply:
Actually if I understand correctly now better, no matter what we have an ordering problem as it seems that all fragments are collected after all fields